### PR TITLE
Determine whether it is necessary to create a service

### LINKED ISSUE #327: [BUG]failed to create service
**Describe the bug**

I created ApplicationList from dashboard.  but failed to create service.

the vela runtime log:
```
2020-09-25T10:05:19.506Z	ERROR	Containerized	Failed to apply a service	{"containerized": "default/test2222", "error": "Service \"test2222\" is invalid: spec.ports: Required value"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128
github.com/oam-dev/kubevela/pkg/controller/v1alpha1/containerized.(*Reconciler).Reconcile
	/workspace/pkg/controller/v1alpha1/containerized/containerized_controller.go:122
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:256
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
```

kubectl get components test2222 -o yaml

```
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  annotations:
    workload.oam.dev/name: backend
  creationTimestamp: "2020-09-25T10:05:08Z"
  generation: 1
  managedFields:
  - apiVersion: core.oam.dev/v1alpha2
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:workload.oam.dev/name: {}
      f:spec:
        .: {}
        f:workload:
          .: {}
          f:apiVersion: {}
          f:kind: {}
          f:metadata:
            .: {}
            f:name: {}
          f:spec:
            .: {}
            f:podSpec:
              .: {}
              f:containers: {}
            f:replicas: {}
    manager: vela
    operation: Update
    time: "2020-09-25T10:05:08Z"
  - apiVersion: core.oam.dev/v1alpha2
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        .: {}
        f:latestRevision:
          .: {}
          f:name: {}
          f:revision: {}
    manager: manager
    operation: Update
    time: "2020-09-25T10:05:09Z"
  name: test2222
  namespace: default
  resourceVersion: "5751130"
  selfLink: /apis/core.oam.dev/v1alpha2/namespaces/default/components/test2222
  uid: 11ad79d2-5360-4e6b-9b8a-282336fbfe78
spec:
  workload:
    apiVersion: standard.oam.dev/v1alpha1
    kind: Containerized
    metadata:
      name: test2222
    spec:
      podSpec:
        containers:
        - image: nginx
          name: test2222
      replicas: 1
status:
  latestRevision:
    name: test2222-v1
    revision: 1
```

**To Reproduce**

1. create a application from dashboard.
2.`kubectl logs -f deployment.apps/vela-core -n oam-system --tail=100` see the service create failed log.

**Expected behavior**

service create success?

**Screenshots**

**Cluster information**

kubernetes version:
v1.19.0

vela version:
```
Version: refs/tags/v0.0.7
GitRevision: git-525c228
GolangVersion: go1.13.15
```

**Additional context**

### PR DESCRIPTION
**Pull request by silenceper**
fix #327 Add not creating service when ports are not set.

### CODE REVIEW

File: pkg/controller/v1alpha1/containerized/containerized_controller.go [hunk 1]
```diff
@@ -117,17 +117,19 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		r.record.Event(eventObj, event.Normal("Service created",
+			fmt.Sprintf("Workload `%s` successfully server side patched a service `%s`",
+				workload.Name, service.Name)))
 	}
-	r.record.Event(eventObj, event.Normal("Service created",
-		fmt.Sprintf("Workload `%s` successfully server side patched a service `%s`",
-			workload.Name, service.Name)))
-
 	// record the new deployment, new service
 	workload.Status.Resources = []cpv1alpha1.TypedReference{
```
Comment (wonderflow) [inline pkg/controller/v1alpha1/containerized/containerized_controller.go:134]: The status should also not contain service object

File: pkg/controller/v1alpha1/containerized/containerized_controller.go [hunk 2]
```diff
@@ -117,17 +117,19 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return util.ReconcileWaitResult,
 			util.PatchCondition(ctx, r, &workload, cpv1alpha1.ReconcileError(errors.Wrap(err, errRenderService)))
 	}
-	// server side apply the service
-	if err := r.Patch(ctx, service, client.Apply, applyOpts...); err != nil {
-		log.Error(err, "Failed to apply a service")
-		r.record.Event(eventObj, event.Warning(errApplyDeployment, err))
-		return util.ReconcileWaitResult,
-			util.PatchCondition(ctx, r, &workload, cpv1alpha1.ReconcileError(errors.Wrap(err, errApplyService)))
+	// Determine whether it is necessary to create a service.
```
Comment (wonderflow) [inline pkg/controller/v1alpha1/containerized/containerized_controller.go:120]: If port not exist, we even shouldn't renderService

Comment (silenceper) [reply pkg/controller/v1alpha1/containerized/containerized_controller.go]: fixed

File: pkg/controller/v1alpha1/containerized/containerized_controller.go (hunk 1 above)
Comment (silenceper) [reply pkg/controller/v1alpha1/containerized/containerized_controller.go]: fixed

**Comment by silenceper** [comment]
add `checkContainerPortsSpecified` function

File: pkg/controller/v1alpha1/containerized/containerized_controller.go [hunk 3]
```diff
@@ -204,6 +210,17 @@ func (r *Reconciler) renderDeployment(ctx context.Context,
 	return deploy, nil
 }
 
+// check whether the container port is specified
+func (r *Reconciler) checkContainerPortsSpecified(ctx context.Context,
+	workload *v1alpha1.Containerized) bool {
```
Comment (wonderflow) [inline pkg/controller/v1alpha1/containerized/containerized_controller.go:215]: check if workload is nil

**Review by wonderflow — APPROVED** [review]
Good job！

File: pkg/controller/v1alpha1/containerized/containerized_controller.go (hunk 3 above)
Comment (silenceper) [reply pkg/controller/v1alpha1/containerized/containerized_controller.go:215]: added

**Pull request merged**